src/components/Notifications/EventLog/EventLogTreeFilter.tsx

```diff
@@ -0,0 +1,211 @@
+import { TreeViewCheckProps } from '@patternfly/react-core/dist/esm/components/TreeView/TreeViewListItem';
+import { AngleDownIcon } from '@patternfly/react-icons';
+import produce from 'immer';
+import React, { ChangeEvent } from 'react';
+
+import { Schemas } from '../../../generated/OpenapiNotifications';
+import { Modify } from '../../../types/Modify';
+import { areEqual } from '../../../utils/Arrays';
+import { EventLogCustomFilter } from './usePrimaryToolbarFilterConfigWrapper';
+
+interface EventLogTreeFilterProps {
+    groups: readonly Schemas.Facet[]
+    placeholder: string
+    filters: EventLogCustomFilter[]
+    updateFilters: React.Dispatch<React.SetStateAction<EventLogCustomFilter[]>>
+}
+
+type TreeNodeItem = Modify<TreeViewDataItem, {
+    id: string,
+    checkProps: TreeViewCheckProps,
+    children?: TreeNodeItem[] | undefined,
+}>
+
+interface TreeNodeDict { [key: string]: TreeNodeItem }
+
+const isChecked = (treeNode: TreeNodeItem) => {
+    return !!treeNode.checkProps.checked;
+};
+
+const childChecked = (treeNode: TreeNodeItem): boolean => {
+    return treeNode.children ? treeNode.children.some(child => childChecked(child)) : isChecked(treeNode);
+};
+
+const allChildrenChecked = (treeNode: TreeNodeItem): boolean  => {
+    return treeNode.children ? treeNode.children.every(child => allChildrenChecked(child)) : isChecked(treeNode);
+};
+
+const initTreeNodeById = (groups: readonly Schemas.Facet[], filters: EventLogCustomFilter[]) => {
+    const init: TreeNodeDict = {};
+    groups.forEach(group => {
+        const currentFilter = filters.find(filter => filter.bundleId === group.name);
+        const currentFilterChipValues = currentFilter?.chips?.map(chip => chip.value);
+
+        const items = group.children as Schemas.Facet[];
+        const checkAll = items.length !== 0 ? items.every(item => currentFilterChipValues?.includes(item.name)) : false;
+        init[group.name] = {
+            id: group.name,
+            name: group.displayName,
+            checkProps: { checked: checkAll || (!currentFilter ? false : null) },
+            children: items.length !== 0 ? items.map(item => ({
+                id: item.name,
+                name: item.displayName,
+                checkProps: { checked: checkAll || currentFilterChipValues?.includes(item.name) }
+            })) : undefined
+        };
+    });
+
+    return init;
+};
+
+export const EventLogTreeFilter: React.FunctionComponent<EventLogTreeFilterProps> = (props) => {
+    const { groups, placeholder, filters, updateFilters } = props;
+
+    const initialize = React.useMemo(() => initTreeNodeById(groups, filters), [ groups, filters ]);
+
+    const [ treeNodeById, setTreeNodeById ] = React.useState<TreeNodeDict>(initialize);
+    const [ isToggled, setIsToggled ] = React.useState(false);
+
+    const treeDataArray = React.useMemo(() => treeNodeById ? Object.values(treeNodeById) : [], [ treeNodeById ]);
```

Comment:
Something to consider. I see that this `treeDataArray` is used to only generate `flattenTree`  which is then used to only generate a `activeParentFilters`  which is then used to generate a `newCustomFilters`. 

I guess we could simplify all that into a single step, and the `newCustomFilters` could be stable enough to do a ref check (i.e. old == new)

Reply:
I use `treeDataArray` also as the data to pass into the props for `TreeView` component... I think having that consistently memoized is beneficial.

I removed the need for flattenTree, so now instead of the effect depending on `flattenTree` it depends on `treeDataArray` which simplifies the logic a little bit. Is your comment referring to a method to remove the data altogether because I'm not too sure that is possible with the current design.

Reply:
I'm not asking to remove, just to condense into a single memo when it makes sense. 
I think it would be worth to take this out and put in a memo (again, it might be condensed with a previous step)
https://github.com/RedHatInsights/notifications-frontend/pull/259/files#diff-b347d923933f0f81d5d65b68fb37caf1d7070f3d58871e06d86144ea3823eec1R126

Reply:
That is a good suggestion, I could extrapolate that into a memo since it only depends on treeDataArray